client/matching/partition_config_provider.go

```diff
@@ -0,0 +1,179 @@
+		// UpdatePartitionConfig updates the partition configuration for a task list
+		UpdatePartitionConfig(domainID string, taskList types.TaskList, taskListType int, config *types.TaskListPartitionConfig)
+	}
+
+	syncedTaskListPartitionConfig struct {
+		sync.RWMutex
+		types.TaskListPartitionConfig
+	}
+
+	partitionConfigProviderImpl struct {
+		configCache         cache.Cache
+		logger              log.Logger
+		domainIDToName      func(string) (string, error)
+		enableReadFromCache dynamicconfig.BoolPropertyFnWithTaskListInfoFilters
+		nReadPartitions     dynamicconfig.IntPropertyFnWithTaskListInfoFilters
+		nWritePartitions    dynamicconfig.IntPropertyFnWithTaskListInfoFilters
+	}
+)
+
+func (c *syncedTaskListPartitionConfig) updateConfig(newConfig types.TaskListPartitionConfig) {
+	c.Lock()
+	defer c.Unlock()
+	if c.Version < newConfig.Version {
+		c.TaskListPartitionConfig = newConfig
+	}
+}
+
+func NewPartitionConfigProvider(
+	logger log.Logger,
+	domainIDToName func(string) (string, error),
+	dc *dynamicconfig.Collection,
+) PartitionConfigProvider {
+	return &partitionConfigProviderImpl{
+		logger:              logger,
+		domainIDToName:      domainIDToName,
+		enableReadFromCache: dc.GetBoolPropertyFilteredByTaskListInfo(dynamicconfig.MatchingEnableGetNumberOfPartitionsFromCache),
+		nReadPartitions:     dc.GetIntPropertyFilteredByTaskListInfo(dynamicconfig.MatchingNumTasklistReadPartitions),
+		nWritePartitions:    dc.GetIntPropertyFilteredByTaskListInfo(dynamicconfig.MatchingNumTasklistWritePartitions),
+		configCache: cache.New(&cache.Options{
+			TTL:             0,
+			InitialCapacity: 100,
+			Pin:             false,
+			MaxCount:        3000,
+			ActivelyEvict:   false,
+		}),
+	}
+}
+
+func (p *partitionConfigProviderImpl) GetNumberOfReadPartitions(domainID string, taskList types.TaskList, taskListType int) int {
+	domainName, err := p.domainIDToName(domainID)
+	if err != nil {
+		return 1
+	}
+	if !p.enableReadFromCache(domainName, taskList.GetName(), taskListType) {
+		return p.nReadPartitions(domainName, taskList.GetName(), taskListType)
+	}
+	c := p.getPartitionConfig(domainID, taskList, taskListType)
+	if c == nil {
+		return 1
+	}
+	c.RLock()
+	defer c.RUnlock()
+	return int(c.NumReadPartitions)
+}
+
+func (p *partitionConfigProviderImpl) GetNumberOfWritePartitions(domainID string, taskList types.TaskList, taskListType int) int {
+	domainName, err := p.domainIDToName(domainID)
+	if err != nil {
+		return 1
+	}
+	if !p.enableReadFromCache(domainName, taskList.GetName(), taskListType) {
+		nPartitions := p.nWritePartitions(domainName, taskList.GetName(), taskListType)
+
+		// checks to make sure number of writes never exceeds number of reads
+		if nRead := p.nReadPartitions(domainName, taskList.GetName(), taskListType); nPartitions > nRead {
+			nPartitions = nRead
+		}
+		return nPartitions
+	}
+	c := p.getPartitionConfig(domainID, taskList, taskListType)
+	if c == nil {
+		return 1
```

Comment:
why falling back to 1 here? config might be evicted from cache which require re-reading it from source of truth. we can't fallback to hardcoded 1.

Reply:
The design is that the first request will always be sent to the root partition and update the cache with the partition config in the response.

Reply:
ok in that case let's add logs to cover what is happening on client side. currently there's no visibility into why it decides what it decides

Reply:
I've added logs to the helper function.
I'm planning to add metrics in another PR and use the metrics to set up alerts.